File: tests/worker/test_sandbox_handoff.py
"""沙箱归属：Gateway 建的沙箱，worker 取产物要用它、任务收尾要还它。

冻结的 ToolGateway 协议只有 catalog / call，沙箱归属没进协议 —— Gateway 按
task_id 自己记着容器，`Task.sandbox_id` 记的则是 worker 兜底建的那种。两边不通
气的话会出两个后果，这个文件把它们各钉一条：

1. **产物丢**：run_python 把 /work/out.png 写在 Gateway 那个沙箱里，worker 取产物
   时看 `task.sandbox_id is None`，自己 acquire 一个全新的空容器，产物必然找不到。
2. **容器泄漏**：Gateway 手上那个沙箱没人 release，只能等 reaper 的 idle_sec
   空闲超时才被收 —— 任务早结束了还占着内存和 CPU 配额。
"""
from worker_fakes import FakePlatform, ScriptedModel, final_turn, make_event, tool_turn

from aite.contracts import TaskStatus
from aite.control import InProcessControlPlane
from aite.evidence import FileEvidenceWriter

PNG = b"\x89PNG\r\n\x1a\n" + b"fake"


# ---- 取产物 ---------------------------------------------------------------


async def test_artifact_is_fetched_from_the_gateway_sandbox(run_task, sandbox, gateway, platform):
    """产物在 Gateway 那个沙箱里 —— worker 得问它要，不能自己另开一个。"""
    sandbox.files["/work/out.png"] = PNG
    script = [
        tool_turn(("run_python", {"code": "plt.savefig('/work/out.png')"})),
        final_turn("图在这里。", [{"path": "/work/out.png", "title": "月度趋势"}]),
    ]

    _, task, _ = await run_task(script, step_seconds=0.6)

    held = "sb_gateway"
    assert [sid for sid, _ in sandbox.get_file_calls] == [held]   # 取的是 Gateway 那个
    assert sandbox.acquired == []                                 # worker 没有另开
    assert len(platform.files) == 1 and platform.files[0].data == PNG
    assert task.status is TaskStatus.delivered


async def test_worker_still_acquires_when_gateway_holds_nothing(run_task, sandbox, platform):
    """兜底路径不变：Gateway 没建过沙箱（比如产物是别处放的），worker 照旧自己建。"""
    sandbox.files["/work/out.txt"] = b"hi"
    script = [final_turn("给你。", [{"path": "/work/out.txt", "title": "结果"}])]

    _, task, _ = await run_task(script)

    assert len(sandbox.acquired) == 1
    assert [sid for sid, _ in sandbox.get_file_calls] == sandbox.acquired
    assert len(platform.files) == 1
    assert task.status is TaskStatus.delivered


# ---- 还沙箱 ---------------------------------------------------------------


async def test_delivered_task_returns_the_gateway_sandbox(run_task, sandbox, gateway):
    """任务交付完，Gateway 的沙箱当场还掉，不等 reaper。"""
    sandbox.files["/work/out.png"] = PNG
    script = [
        tool_turn(("run_python", {"code": "plt.savefig('/work/out.png')"})),
        final_turn("图在这里。", [{"path": "/work/out.png", "title": "月度趋势"}]),
    ]

    _, task, _ = await run_task(script, step_seconds=0.6)

    assert task.status is TaskStatus.delivered
    assert gateway.released_tasks == [task.id]
    assert sandbox.released == ["sb_gateway"]
    assert gateway.sandbox_id_of(task.id) is None


async def test_failed_task_returns_the_gateway_sandbox(run_task, sandbox, gateway):
    """failed 也走同一个收尾口，沙箱一样要还。"""
    script = [tool_turn(("run_python", {"code": "1/0"}))] * 3
    _, task, _ = await run_task(script, repeat=tool_turn(("run_python", {"code": "1/0"})))

    assert task.status is TaskStatus.failed        # 撞步数上限
    assert gateway.released_tasks == [task.id]
    assert sandbox.released == ["sb_gateway"]


async def test_cancel_mid_run_returns_the_gateway_sandbox(store, config, clock, sandbox, gateway):
    """跑到一半被 !stop：任务在 worker 手里，收尾由 worker 做，沙箱在那里还。"""
    platform = FakePlatform()
    plane: InProcessControlPlane | None = None
    task_id: str | None = None

    async def inject(step: int) -> None:
        if step == 1:                   # 第一步（run_python）已经建了沙箱，这时点 stop
            await plane.cancel_task(await store.get_task(task_id))

    model = ScriptedModel(
        [
            tool_turn(("run_python", {"code": "长活"})),
            tool_turn(("checklist_note", {"text": "还在算"})),   # 跑完这步回到循环开头才检查取消
            final_turn("不该走到这一步"),
        ],
        clock=clock,
        step_seconds=0.6,
        on_call=inject,
    )
    evidence = FileEvidenceWriter(config.storage.evidence_dir)
    plane = InProcessControlPlane(
        store=store, platform=platform, evidence=evidence, config=config,
        model=model, gateway=gateway, sandbox=sandbox, clock=clock, sleep=clock.sleep,
    )

    await plane.handle_event(make_event(text="跑个长活"))
    task_id = (await store.list_active_tasks("oc_chat"))[0].id
    await plane.run_pending()

    task = await store.get_task(task_id)
    assert task.status is TaskStatus.cancelled
    assert gateway.released_tasks == [task_id]
    assert sandbox.released == ["sb_gateway"]


async def test_cancel_before_worker_starts_returns_the_gateway_sandbox(
    store, config, clock, sandbox, gateway, platform
):
    """任务还没被 worker 领走就被停：worker 不会再收尾，得由控制面还沙箱。"""
    evidence = FileEvidenceWriter(config.storage.evidence_dir)
    plane = InProcessControlPlane(
        store=store, platform=platform, evidence=evidence, config=config,
        model=ScriptedModel([final_turn("不该跑到这")], clock=clock),
        gateway=gateway, sandbox=sandbox, clock=clock, sleep=clock.sleep,
    )

    await plane.handle_event(make_event(text="先建个任务"))
    task = (await store.list_active_tasks("oc_chat"))[0]
    gateway.hold_sandbox(task.id)          # 演上一轮工具调用留下的沙箱

    await plane.cancel_task(task)          # 还没 run_pending，任务不在 worker 手里

    assert (await store.get_task(task.id)).status is TaskStatus.cancelled
    assert gateway.released_tasks == [task.id]
    assert sandbox.released == ["sb_gateway"]

File: core/crates/worker/tests/test_sandbox_handoff.rs
//! 沙箱归属：Gateway 建的沙箱，worker 取产物要用它、任务收尾要还它。
//! 移植自 `tests/worker/test_sandbox_handoff.py`（6 条）。
//!
//! 旧契约的 `ToolGateway` 只有 catalog / call，沙箱归属没进协议 —— Gateway 按 task_id
//! 自己记着容器，`Task.sandbox_id` 记的则是 worker 兜底建的那种。p0.2 把
//! `sandbox_id_of / release_task` 显式化了（§3.1 差异表 D2），但两边不通气的后果不变：
//!
//! 1. **产物丢**：run_python 把 /work/out.png 写在 Gateway 那个沙箱里，worker 取产物时
//!    看 `task.sandbox_id is None`，自己 acquire 一个全新的空容器，产物必然找不到。
//! 2. **容器泄漏**：Gateway 手上那个沙箱没人 release，只能等 reaper 的 idle_sec 空闲超时
//!    才被收 —— 任务早结束了还占着内存和 CPU 配额。
mod common;

use aite_contracts::TaskStatus;
use common::*;
use serde_json::json;
use std::sync::Arc;

const PNG: &[u8] = b"\x89PNG\r\n\x1a\nfake";

fn chart_script() -> Vec<aite_contracts::ModelTurn> {
    vec![
        tool_turn(&[(
            "run_python",
            json!({"code": "plt.savefig('/work/out.png')"}),
        )]),
        final_turn_with(
            "图在这里。",
            json!([{"path": "/work/out.png", "title": "月度趋势"}]),
        ),
    ]
}

// ---- 取产物 -------------------------------------------------------------

#[tokio::test]
async fn artifact_is_fetched_from_the_gateway_sandbox() {
    // 产物在 Gateway 那个沙箱里 —— worker 得问它要，不能自己另开一个
    let mut h = Harness::new();
    h.sandbox.put("/work/out.png", PNG);
    h.seed("帮我出个图", Vec::new()).await;
    let model = Arc::new(ScriptedModel::new(chart_script()).with_clock(h.clock.clone(), 0.6));
    let task = h.run(model).await;

    let used: Vec<String> = h
        .sandbox
        .get_file_calls()
        .into_iter()
        .map(|(s, _)| s)
        .collect();
    assert_eq!(used, ["sb_gateway"], "取的是 Gateway 那个");
    assert!(h.sandbox.acquired().is_empty(), "worker 没有另开");
    assert_eq!(h.platform.files().len(), 1);
    assert_eq!(h.platform.files()[0].data, PNG);
    assert_eq!(task.status, TaskStatus::Delivered);
}

#[tokio::test]
async fn worker_still_acquires_when_gateway_holds_nothing() {
    // 兜底路径不变：Gateway 没建过沙箱（比如产物是别处放的），worker 照旧自己建
    let mut h = Harness::new();
    h.sandbox.put("/work/out.txt", b"hi");
    h.seed("帮我出个图", Vec::new()).await;
    let model = Arc::new(ScriptedModel::new(vec![final_turn_with(
        "给你。",
        json!([{"path": "/work/out.txt", "title": "结果"}]),
    )]));
    let task = h.run(model).await;

    assert_eq!(h.sandbox.acquired().len(), 1);
    let used: Vec<String> = h
        .sandbox
        .get_file_calls()
        .into_iter()
        .map(|(s, _)| s)
        .collect();
    assert_eq!(used, h.sandbox.acquired());
    assert_eq!(h.platform.files().len(), 1);
    assert_eq!(task.status, TaskStatus::Delivered);
}

// ---- 还沙箱 -------------------------------------------------------------

#[tokio::test]
async fn delivered_task_returns_the_gateway_sandbox() {
    // 任务交付完，Gateway 的沙箱当场还掉，不等 reaper
    let mut h = Harness::new();
    h.sandbox.put("/work/out.png", PNG);
    h.seed("帮我出个图", Vec::new()).await;
    let model = Arc::new(ScriptedModel::new(chart_script()).with_clock(h.clock.clone(), 0.6));
    let task = h.run(model).await;

    assert_eq!(task.status, TaskStatus::Delivered);
    assert_eq!(h.gateway.released_tasks(), vec![task.id.clone()]);
    assert_eq!(h.sandbox.released(), ["sb_gateway"]);
    assert_eq!(
        aite_contracts::ToolGateway::sandbox_id_of(&*h.gateway, &task.id).await,
        None
    );
}

#[tokio::test]
async fn failed_task_returns_the_gateway_sandbox() {
    // failed 也走同一个收尾口，沙箱一样要还（这里由 T20 的第 5 次重复开火）
    let run = run_repeat(tool_turn(&[("run_python", json!({"code": "1/0"}))]), 0.6).await;

    assert_eq!(run.task.status, TaskStatus::Failed);
    assert_eq!(run.h.gateway.released_tasks(), vec![run.task.id.clone()]);
    assert_eq!(run.h.sandbox.released(), ["sb_gateway"]);
}

#[tokio::test]
async fn cancel_mid_run_returns_the_gateway_sandbox() {
    // 跑到一半被 !stop：任务在 worker 手里，收尾由 worker 做，沙箱在那里还
    let mut h = Harness::new();
    h.seed("跑个长活", Vec::new()).await;
    let h = Arc::new(h);
    let hook_h = h.clone();
    let model = Arc::new(
        ScriptedModel::new(vec![
            tool_turn(&[("run_python", json!({"code": "长活"}))]),
            // 跑完这步回到循环开头才检查取消
            tool_turn(&[("checklist_note", json!({"text": "还在算"}))]),
            final_turn("不该走到这一步"),
        ])
        .with_clock(h.clock.clone(), 0.6)
        .with_on_call(Arc::new(move |step| {
            let h = hook_h.clone();
            Box::pin(async move {
                // 第一步（run_python）已经建了沙箱，这时点 stop
                if step == 1 {
                    h.cancel();
                }
            })
        })),
    );
    let task = h.run(model).await;

    assert_eq!(task.status, TaskStatus::Cancelled);
    assert_eq!(h.gateway.released_tasks(), vec![task.id.clone()]);
    assert_eq!(h.sandbox.released(), ["sb_gateway"]);
}

#[tokio::test]
async fn cancel_before_the_first_step_returns_the_gateway_sandbox() {
    // Python 那条走的是控制面「任务还没被 worker 领走」的分支（沙箱由控制面还，归 R4）。
    // worker 这一侧对应的不变量是：取消旗在第一步之前就竖着时，模型一次都不调，
    // 但 cancelled 证据、卡片与沙箱归还一样都不少。
    let mut h = Harness::new();
    h.seed("先建个任务", Vec::new()).await;
    h.gateway.hold_sandbox(&h.task.id); // 演上一轮工具调用留下的沙箱
    h.cancel();
    let model = Arc::new(ScriptedModel::new(vec![final_turn("不该跑到这")]));
    let task = h.run(model.clone()).await;

    assert_eq!(model.call_count(), 0);
    assert_eq!(task.status, TaskStatus::Cancelled);
    assert!(h.platform.cards().is_empty(), "没跑到发卡片那一步");
    assert_eq!(h.gateway.released_tasks(), vec![task.id.clone()]);
    assert_eq!(h.sandbox.released(), ["sb_gateway"]);
}
